File: .claude/skills/startup-multi-device-analysis/scripts/variance_analysis.py
#!/usr/bin/env python3
"""Per-iteration variance analysis over macrobenchmark .perfetto-trace files.

Companion to variance_metrics.sql (same directory); produced the dataset in
startup-variance-analysis-2026-08-11.txt. Usage:
  python3 variance_analysis.py <trace_processor> <traces-dir> [--json PATH] [--out PATH]
      [--little-cpus CPU,CPU,...]

--little-cpus sets the cpu ids that make up the little cluster for the section-E residency
split (default "0,1,2,3"); pass the little_cpus value from device_probe.py's topology JSON
for the target device rather than assuming the default.

--json dumps the raw per-iteration dataset (window, section durations, per-CPU residency,
thread states) for downstream aggregation; --out tees the printed report to a file.

Prints:
  A. per-iteration matrix: window + canonical section durations (spot outliers/trends)
  B. per-section fluctuation stats: median, max, spread, stdev, Pearson r vs window
  C. outlier decomposition: for the slowest iterations, which sections exceed their medians
  D. per-section thread-state split (Running vs blocked S/D/IO), median vs slowest iterations
  E. per-cluster core residency of the main thread inside the window
"""

import argparse
import contextlib
import csv
import io
import json
import os
import re
import statistics
import subprocess
import sys

CANONICAL = [
    "emb-embrace-impl-init",
    "emb-bootstrapper-init",
    "emb-modules-init",
    "emb-persisted-config-load",
    "emb-config-service-init",
    "emb-span-service-init",
    "emb-otel-tracer-init",
    "emb-essential-service-init",
    "emb-delivery-init",
    "emb-payload-source-init",
    "emb-post-init",
    "emb-post-services-setup",
    "emb-load-instrumentation",
]
EXTRAS = [
    "emb-install-native-crash-signal-handlers",
    "emb-load-embrace-native-lib",
    "emb-record-startup",
    "emb-power-service-registration",
    "emb-snapshot-session",
]
SHORT = {
    "emb-embrace-impl-init": "impl",
    "emb-bootstrapper-init": "boot",
    "emb-modules-init": "modules",
    "emb-persisted-config-load": "cfgload",
    "emb-config-service-init": "cfgsvc",
    "emb-span-service-init": "spansvc",
    "emb-otel-tracer-init": "tracer",
    "emb-essential-service-init": "essent",
    "emb-delivery-init": "deliv",
    "emb-payload-source-init": "payload",
    "emb-post-init": "postini",
    "emb-post-services-setup": "postsvc",
    "emb-load-instrumentation": "loadins",
    "emb-install-native-crash-signal-handlers": "sighand",
    "emb-load-embrace-native-lib": "natlib",
    "emb-record-startup": "recstart",
    "emb-power-service-registration": "power",
    "emb-snapshot-session": "snap",
}


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("tp")
    ap.add_argument("traces_dir")
    ap.add_argument("--json", dest="json_path")
    ap.add_argument("--out", dest="out_path")
    ap.add_argument("--little-cpus", dest="little_cpus", default="0,1,2,3",
                     help="comma-separated cpu ids for the little cluster "
                          "(see device_probe.py's little_cpus output)")
    args = ap.parse_args()
    little_cpus = {int(x) for x in args.little_cpus.split(",")}

    sql = os.path.join(os.path.dirname(os.path.abspath(__file__)), "variance_metrics.sql")
    traces = sorted(
        (f for f in os.listdir(args.traces_dir) if f.endswith(".perfetto-trace")),
        key=lambda f: int(re.search(r"iter(\d+)", f).group(1)),
    )
    data = []
    for name in traces:
        data.append((name, extract(args.tp, sql, os.path.join(args.traces_dir, name))))

    if args.json_path:
        with open(args.json_path, "w") as f:
            json.dump([{"trace": name, **m} for name, m in data], f)

    buf = io.StringIO()
    with contextlib.redirect_stdout(buf):
        report(data, little_cpus)
    sys.stdout.write(buf.getvalue())
    if args.out_path:
        with open(args.out_path, "w") as f:
            f.write(buf.getvalue())
    return 0


def report(data, little_cpus) -> None:
    n = len(data)
    windows = [m["window_ms"] for _, m in data]
    wmed = statistics.median(windows)

    # A: per-iteration matrix
    cols = CANONICAL + EXTRAS
    print("A. per-iteration durations, ms")
    print("  it   window " + " ".join(f"{SHORT[c]:>8}" for c in cols))
    for i, (_, m) in enumerate(data):
        row = " ".join(f"{m['dur'].get(c, float('nan')):>8.2f}" for c in cols)
        print(f"  {i:>3} {m['window_ms']:>8.1f} {row}")
    print()

    # B: fluctuation stats + correlation with window
    print("B. per-section fluctuation (n=%d), ms" % n)
    print(f"  {'section':<42}{'min':>7}{'med':>7}{'max':>7}{'spread':>8}{'stdev':>7}{'r_win':>7}")
    for c in cols:
        vals = [m["dur"].get(c) for _, m in data]
        vals = [v for v in vals if v is not None]
        if len(vals) < n:
            continue
        r = pearson(vals, windows)
        print(f"  {c:<42}{min(vals):>7.2f}{statistics.median(vals):>7.2f}{max(vals):>7.2f}"
              f"{max(vals) - min(vals):>8.2f}{statistics.stdev(vals):>7.2f}{r:>7.2f}")
    print()

    # C: outlier decomposition (window sections only; parents include children)
    med = {c: statistics.median([m["dur"][c] for _, m in data if c in m["dur"]]) for c in cols}
    slow = sorted(range(n), key=lambda i: -windows[i])[:4]
    print("C. slowest iterations: section excess vs median (window Δ = window - median window)")
    for i in slow:
        m = data[i][1]
        deltas = sorted(
            ((c, m["dur"].get(c, 0) - med[c]) for c in cols),
            key=lambda kv: -kv[1],
        )[:7]
        ds = ", ".join(f"{SHORT[c]} +{d:.1f}" for c, d in deltas if d > 0.3)
        print(f"  iter{i:03d}  window {windows[i]:.1f} (Δ +{windows[i] - wmed:.1f}): {ds}")
    print()

    # D: thread-state split per section — median vs the slowest iterations
    print("D. thread-state inside each section, ms (median across iters | mean over 4 slowest iters)")
    print(f"  {'section':<42}{'run':>7}{'sleep':>7}{'io/D':>7}{'rq':>6}  |"
          f"{'run':>7}{'sleep':>7}{'io/D':>7}{'rq':>6}")
    for c in cols:
        med_split = split_stats(data, c, range(n))
        slow_split = split_stats(data, c, slow)
        if med_split is None or slow_split is None:
            continue
        print(f"  {c:<42}" + "".join(f"{v:>7.2f}" for v in med_split[:3]) + f"{med_split[3]:>6.2f}"
              + "  |" + "".join(f"{v:>7.2f}" for v in slow_split[:3]) + f"{slow_split[3]:>6.2f}")
    print()

    # E: per-cluster residency inside window (SM-A145M: cluster0 = cpu0-3, cluster1 = cpu4-7,
    # identical A55 cores, independent cpufreq policies)
    cpus = sorted({int(k) for _, m in data for k in m["cpu_ms"]})
    print("E. main-thread CPU residency inside window, ms per cpu")
    print("  it   window " + " ".join(f"cpu{c:>4}" for c in cpus) + "   little-share")
    shares = []
    for i, (_, m) in enumerate(data):
        tot = sum(m["cpu_ms"].values()) or 1.0
        cl0 = sum(v for k, v in m["cpu_ms"].items() if int(k) in little_cpus) / tot
        shares.append(cl0)
        row = " ".join(f"{m['cpu_ms'].get(str(c), 0):>7.1f}" for c in cpus)
        print(f"  {i:>3} {m['window_ms']:>8.1f} {row}   {cl0 * 100:>5.1f}%")
    print(f"  r(window, little-share) = {pearson(windows, shares):.2f}")
    cl0_wins = [w for w, s in zip(windows, shares) if s > 0.5]
    cl1_wins = [w for w, s in zip(windows, shares) if s <= 0.5]
    if cl0_wins and cl1_wins:
        print(f"  little-majority: n={len(cl0_wins)} mean={statistics.mean(cl0_wins):.1f}"
              f"  |  cluster1-majority: n={len(cl1_wins)} mean={statistics.mean(cl1_wins):.1f}")


def split_stats(data, section, idxs):
    runs, sleeps, ios, rqs = [], [], [], []
    for i in idxs:
        m = data[i][1]
        if section not in m["dur"]:
            return None
        st = m["states"].get(section, {})
        runs.append(st.get("Running", 0))
        sleeps.append(st.get("S", 0))
        ios.append(st.get("D", 0) + st.get("D+io", 0) + st.get("DK", 0) + st.get("DK+io", 0)
                   + st.get("S+io", 0))
        rqs.append(st.get("R", 0) + st.get("R+", 0))
    agg = statistics.median if len(list(idxs)) > 6 else statistics.mean
    return (agg(runs), agg(sleeps), agg(ios), agg(rqs))


def extract(tp, sql, trace_path):
    cmd = [tp, "-q", sql, trace_path]
    if not os.access(tp, os.X_OK):
        cmd = [sys.executable] + cmd
    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise RuntimeError(f"trace_processor failed on {trace_path}:\n{proc.stderr}")
    m = {"dur": {}, "cpu_ms": {}, "states": {}}
    seen = False
    for row in csv.reader(io.StringIO(proc.stdout)):
        if row == ["what", "k", "val"]:
            seen = True
            continue
        if not seen or len(row) != 3 or row[2] == "[NULL]":
            continue
        what, k, val = row[0], row[1], float(row[2])
        if what == "window_ms":
            m["window_ms"] = val
        elif what == "cpu_ms":
            m["cpu_ms"][k] = val
        elif what == "dur":
            m["dur"][k] = val
        elif what.startswith("st:"):
            m["states"].setdefault(k, {})[what[3:]] = val
    return m


def pearson(xs, ys):
    mx, my = statistics.mean(xs), statistics.mean(ys)
    num = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    den = (sum((x - mx) ** 2 for x in xs) * sum((y - my) ** 2 for y in ys)) ** 0.5
    if den == 0:
        return float("nan")
    return num / den


if __name__ == "__main__":
    sys.exit(main())

File: startup-tools/src/main/kotlin/io/embrace/startup/analysis/VarianceAnalysis.kt
package io.embrace.startup.analysis

import io.embrace.startup.core.stats.Descriptive
import io.embrace.startup.core.stats.Quantile
import io.embrace.startup.core.text.PyFormat
import io.embrace.startup.perfetto.Queries
import io.embrace.startup.perfetto.TraceProcessor
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.nio.file.Path

/**
 * `variance_analysis.py`: per-iteration variance over a directory of traces - the iteration matrix
 * (A), per-section fluctuation and correlation with the window (B), outlier decomposition of the
 * four slowest iterations (C), the thread-state split inside each section (D) and the main thread's
 * per-CPU residency with the little-cluster share (E).
 *
 * The `--json` dataset is the input contract of `hypothesis-tests`, `factors-report` and
 * `cross-device-sections`; [Record] is that schema.
 */
object VarianceAnalysis {

    val CANONICAL: List<String> = listOf(
        "emb-embrace-impl-init", "emb-bootstrapper-init", "emb-modules-init", "emb-persisted-config-load",
        "emb-config-service-init", "emb-span-service-init", "emb-otel-tracer-init", "emb-essential-service-init",
        "emb-delivery-init", "emb-payload-source-init", "emb-post-init", "emb-post-services-setup",
        "emb-load-instrumentation",
    )
    val EXTRAS: List<String> = listOf(
        "emb-install-native-crash-signal-handlers",
        "emb-load-embrace-native-lib",
        "emb-record-startup",
        "emb-power-service-registration",
        "emb-snapshot-session",
    )
    val SHORT: Map<String, String> = mapOf(
        "emb-embrace-impl-init" to "impl",
        "emb-bootstrapper-init" to "boot",
        "emb-modules-init" to "modules",
        "emb-persisted-config-load" to "cfgload",
        "emb-config-service-init" to "cfgsvc",
        "emb-span-service-init" to "spansvc",
        "emb-otel-tracer-init" to "tracer",
        "emb-essential-service-init" to "essent",
        "emb-delivery-init" to "deliv",
        "emb-payload-source-init" to "payload",
        "emb-post-init" to "postini",
        "emb-post-services-setup" to "postsvc",
        "emb-load-instrumentation" to "loadins",
        "emb-install-native-crash-signal-handlers" to "sighand",
        "emb-load-embrace-native-lib" to "natlib",
        "emb-record-startup" to "recstart",
        "emb-power-service-registration" to "power",
        "emb-snapshot-session" to "snap",
    )
    val DEFAULT_LITTLE_CPUS: Set<Int> = setOf(0, 1, 2, 3)

    /** One trace's dataset - the `--json` element schema, key names as the Python wrote them. */
    @Serializable
    data class Record(
        val trace: String,
        /** Absent when the trace has no `emb-sdk-start` slice; the Python then failed at report time. */
        @SerialName("window_ms") val windowMs: Double? = null,
        /** First-occurrence duration of every `emb-*` section. */
        val dur: Map<String, Double>,
        /** Main-thread residency per CPU (string keys, as JSON has them). */
        @SerialName("cpu_ms") val cpuMs: Map<String, Double>,
        /** Per section, per thread state (`Running`, `S`, `D+io`, ...) ms of the emitting thread. */
        val states: Map<String, Map<String, Double>>,
    )

    fun recordOf(trace: String, triples: List<TraceProcessor.Triple>): Record {
        var window: Double? = null
        val dur = LinkedHashMap<String, Double>()
        val cpu = LinkedHashMap<String, Double>()
        val states = LinkedHashMap<String, LinkedHashMap<String, Double>>()
        triples.forEach { t ->
            val value = t.value?.toDouble() ?: return@forEach
            when {
                t.what == "window_ms" -> window = value
                t.what == "cpu_ms" -> cpu[t.k] = value
                t.what == "dur" -> dur[t.k] = value
                t.what.startsWith("st:") -> states.getOrPut(t.k) { LinkedHashMap() }[t.what.removePrefix("st:")] = value
            }
        }
        return Record(trace, window, dur, cpu, states)
    }

    fun extract(tp: TraceProcessor, trace: Path): Record =
        recordOf(trace.fileName.toString(), tp.triples(Queries.VARIANCE_METRICS, trace))

    fun report(data: List<Record>, littleCpus: Set<Int> = DEFAULT_LITTLE_CPUS): String {
        val out = StringBuilder()
        val windows = data.map { requireNotNull(it.windowMs) { "${it.trace} has no emb-sdk-start window" } }
        val cols = CANONICAL + EXTRAS
        out.matrix(data, windows, cols)
        out.fluctuation(data, windows, cols)
        val slow = windows.indices.sortedByDescending { windows[it] }.take(SLOWEST)
        out.decomposition(data, windows, cols, slow)
        out.threadStates(data, cols, slow)
        out.residency(data, windows, littleCpus)
        return out.toString()
    }

    private fun StringBuilder.line(s: String = ""): StringBuilder = append(s).append('\n')

    private fun StringBuilder.matrix(data: List<Record>, windows: List<Double>, cols: List<String>) {
        line("A. per-iteration durations, ms")
        line("  it   window " + cols.joinToString(" ") { SHORT.getValue(it).padStart(W8) })
        data.forEachIndexed { i, m ->
            val row = cols.joinToString(" ") { PyFormat.fixed(m.dur[it] ?: Double.NaN, 2).padStart(W8) }
            line("  ${i.toString().padStart(W3)} ${PyFormat.fixed(windows[i], 1).padStart(W8)} $row")
        }
        line()
    }

    private fun StringBuilder.fluctuation(data: List<Record>, windows: List<Double>, cols: List<String>) {
        line("B. per-section fluctuation (n=${data.size}), ms")
        line(
            "  ${"section".padEnd(W42)}${"min".padStart(W7)}${"med".padStart(W7)}${"max".padStart(W7)}" +
                "${"spread".padStart(W8)}${"stdev".padStart(W7)}${"r_win".padStart(W7)}",
        )
        cols.forEach { c ->
            val vals = data.mapNotNull { it.dur[c] }
            if (vals.size < data.size) return@forEach
            val sorted = vals.sorted()
            line(
                "  ${c.padEnd(W42)}${f2(sorted.first(), W7)}${f2(Quantile.median(sorted), W7)}${f2(sorted.last(), W7)}" +
                    "${f2(sorted.last() - sorted.first(), W8)}${f2(Descriptive.sampleStdev(vals), W7)}" +
                    f2(Descriptive.pearson(vals, windows), W7),
            )
        }
        line()
    }

    private fun StringBuilder.decomposition(data: List<Record>, windows: List<Double>, cols: List<String>, slow: List<Int>) {
        val wmed = Quantile.median(windows.sorted())
        // The Python's statistics.median raised on a section present in NO trace; NaN here keeps the
        // report renderable and the "> 0.3" filter drops it, which is the only sane reading.
        val med = cols.associateWith { c ->
            val vals = data.mapNotNull { it.dur[c] }
            if (vals.isEmpty()) Double.NaN else Quantile.median(vals.sorted())
        }
        line("C. slowest iterations: section excess vs median (window Δ = window - median window)")
        slow.forEach { i ->
            val m = data[i]
            val deltas = cols.map { c -> c to ((m.dur[c] ?: 0.0) - med.getValue(c)) }
                .sortedByDescending { it.second }
                .take(TOP_DELTAS)
            val ds = deltas.filter { it.second > DELTA_FLOOR }
                .joinToString(", ") { (c, d) -> "${SHORT.getValue(c)} +${PyFormat.fixed(d, 1)}" }
            val iter = i.toString().padStart(ITER_DIGITS, '0')
            line("  iter$iter  window ${PyFormat.fixed(windows[i], 1)} (Δ +${PyFormat.fixed(windows[i] - wmed, 1)}): $ds")
        }
        line()
    }

    private fun StringBuilder.threadStates(data: List<Record>, cols: List<String>, slow: List<Int>) {
        line("D. thread-state inside each section, ms (median across iters | mean over 4 slowest iters)")
        line(
            "  ${"section".padEnd(W42)}${"run".padStart(W7)}${"sleep".padStart(W7)}${"io/D".padStart(W7)}${"rq".padStart(W6)}  |" +
                "${"run".padStart(W7)}${"sleep".padStart(W7)}${"io/D".padStart(W7)}${"rq".padStart(W6)}",
        )
        cols.forEach { c ->
            val medSplit = splitStats(data, c, data.indices.toList()) ?: return@forEach
            val slowSplit = splitStats(data, c, slow) ?: return@forEach
            line("  ${c.padEnd(W42)}${splitCells(medSplit)}  |${splitCells(slowSplit)}")
        }
        line()
    }

    private fun StringBuilder.residency(data: List<Record>, windows: List<Double>, littleCpus: Set<Int>) {
        val cpus = data.flatMap { m -> m.cpuMs.keys.map { it.toInt() } }.toSortedSet().toList()
        line("E. main-thread CPU residency inside window, ms per cpu")
        line("  it   window " + cpus.joinToString(" ") { "cpu${it.toString().padStart(W4)}" } + "   little-share")
        val shares = ArrayList<Double>()
        data.forEachIndexed { i, m ->
            val tot = m.cpuMs.values.sum().takeIf { it != 0.0 } ?: 1.0
            val cl0 = m.cpuMs.filterKeys { it.toInt() in littleCpus }.values.sum() / tot
            shares.add(cl0)
            val row = cpus.joinToString(" ") { PyFormat.fixed(m.cpuMs[it.toString()] ?: 0.0, 1).padStart(W7) }
            line(
                "  ${i.toString().padStart(
                    W3,
                )} ${PyFormat.fixed(windows[i], 1).padStart(W8)} $row   ${PyFormat.fixed(cl0 * PERCENT, 1).padStart(W5)}%",
            )
        }
        line("  r(window, little-share) = ${PyFormat.fixed(Descriptive.pearson(windows, shares), 2)}")
        val cl0Wins = windows.filterIndexed { i, _ -> shares[i] > HALF }
        val cl1Wins = windows.filterIndexed { i, _ -> shares[i] <= HALF }
        if (cl0Wins.isNotEmpty() && cl1Wins.isNotEmpty()) {
            line(
                "  little-majority: n=${cl0Wins.size} mean=${PyFormat.fixed(PyFormat.exactMean(cl0Wins), 1)}" +
                    "  |  cluster1-majority: n=${cl1Wins.size} mean=${PyFormat.fixed(PyFormat.exactMean(cl1Wins), 1)}",
            )
        }
    }

    /** `split_stats`: (run, sleep, io/D, runqueue) per section over the given iterations; null if any lacks the section. */
    private fun splitStats(data: List<Record>, section: String, idxs: List<Int>): List<Double>? {
        val runs = ArrayList<Double>()
        val sleeps = ArrayList<Double>()
        val ios = ArrayList<Double>()
        val rqs = ArrayList<Double>()
        idxs.forEach { i ->
            val m = data[i]
            if (section !in m.dur) return null
            val st = m.states[section] ?: emptyMap()
            runs.add(st["Running"] ?: 0.0)
            sleeps.add(st["S"] ?: 0.0)
            ios.add(
                (st["D"] ?: 0.0) + (st["D+io"] ?: 0.0) + (st["DK"] ?: 0.0) + (st["DK+io"] ?: 0.0) + (st["S+io"] ?: 0.0),
            )
            rqs.add((st["R"] ?: 0.0) + (st["R+"] ?: 0.0))
        }
        val agg: (List<Double>) -> Double = if (idxs.size > MEDIAN_ABOVE) {
            { Quantile.median(it.sorted()) }
        } else {
            { PyFormat.exactMean(it) }
        }
        return listOf(agg(runs), agg(sleeps), agg(ios), agg(rqs))
    }

    private fun splitCells(split: List<Double>): String =
        split.take(3).joinToString("") { f2(it, W7) } + f2(split[3], W6)

    private fun f2(x: Double, width: Int) = PyFormat.fixed(x, 2).padStart(width)

    private const val SLOWEST = 4
    private const val ITER_DIGITS = 3
    private const val TOP_DELTAS = 7
    private const val DELTA_FLOOR = 0.3
    private const val MEDIAN_ABOVE = 6
    private const val HALF = 0.5
    private const val PERCENT = 100.0
    private const val W3 = 3
    private const val W4 = 4
    private const val W5 = 5
    private const val W6 = 6
    private const val W7 = 7
    private const val W8 = 8
    private const val W42 = 42
}
